src/v/cluster/topic_metadata_item.cc

```diff
@@ -10,8 +10,26 @@
  */
 #include "cluster/topic_metadata_item.h"
 
+#include "config/configuration.h"
+#include "prometheus/prometheus_sanitize.h"
+
+#include <seastar/core/metrics.hh>
+#include <seastar/core/metrics_registration.hh>
+
 namespace cluster {
 
+topic_metadata_item::topic_metadata_item(topic_metadata _metadata) noexcept
+  : metadata(std::move(_metadata)) {
+    setup_metrics();
+}
+
+topic_metadata_item::topic_metadata_item(topic_metadata_item&& other) noexcept
+  : metadata(std::move(other.metadata))
+  , replica_revisions(std::move(other.replica_revisions))
+  , _public_metrics(std::move(other._public_metrics)) {
+    setup_metrics();
+}
```

Comment:
1. if immediately after moving _public_metrics setup_metrics calls _public_metrics.clear() why move it at all?
2. is there a concern about a double-registration failure in seastar here? setup metrics will register some metrics but the moved-from object had metrics registered that presumably won't be unregistered until its destructor runs, after the move constructor here.
3. i wonder if it would be cleaner to make topic_metadata_item a move-only type (you might need to adjust the container type hold it) or wrap the item in a unique_ptr

Reply:
1. The `clear` is required because moving a `ss::metrics::metric_groups` object preserves the metrics registered with it. (see 2. too)
2. Yes, there is a concern and this is why this code looks a bit odd. However, moving the `ss::metrics::metric_groups` object moves the registered metrics too (they are kept alive). `setup_metrics` calls `metric_groups::clear()` before registering any new metrics which guards against double registration.
3. `topic_metadata_item` is a move-only type since it holds a `ss::metrics::metric_groups` object which is move-only. I guess I could explicitly mark the copy constructor deleted to make it clearer.

Reply:
I tend to go with rule-of-5 or rule-of-0. A lot of classes in the code base always specify special members as `= default` or `= delete`

Reply:
rule-of-5 where everything except constructor and destroctor are all deleted is pretty common, hence the unique ptr suggestion

Reply:
Thanks for the advice. I went for explicitly defining special members, even if its a bit verbose. `metrics::metric_groups` is already a `unique_ptr` under the hood so I wanted to avoid doing too much pointer chasing.